workflow/controller/dag.go

```diff
@@ -127,6 +127,42 @@ func generatePhaseNodes(children []string, branchPhase wfv1.NodePhase) []phaseNo
 	return out
 }
 
+type uniqueQueue struct {
+	current map[string]bool
+	queue   []phaseNode
+}
+
+func newUniqueQueue() *uniqueQueue {
+	return &uniqueQueue{
+		current: make(map[string]bool),
+		queue:   []phaseNode{},
+	}
+}
+
+func (uq *uniqueQueue) add(nodes ...phaseNode) {
+	for _, node := range nodes {
```

Comment:
I think you need a mutex lock to avoid the concurrent access issue.

Reply:
This runs on the same goroutine